### PR TITLE
Continued webpki CRL support.

### PR DESCRIPTION
**Pull request by cpu**
## Description

This branch adds support for verifying certificates are not revoked by consulting CRLs provided through implementations of a simple provider trait. Trust anchor revocation status is not considered.

In general I've tried to adapt to the existing webpki code without making any significant invasive changes to the path building logic. There's not a lot of test coverage in this area of the codebase and it's a tricky problem domain with many pitfalls.

## Points for discussion

There are some rough edges left to be figured out as follow-up work:

* Error messages for revocation related events are incorrect. All error states are reported as "UnknownIssuer". See https://github.com/rustls/webpki/issues/79

* There is no enforcement of the current time being greater than the CRL's `NextUpdate`. I found most implementations I surveyed chose not to treat this as a hard error so for now I've omitted any enforcement. Maybe this should be configurable? 

* There are some optimizations left on the table. Providing a relevant CRL is handled by a trait that could do some kind of caching if desired, but the actual act of iterating the relevant CRL to find a particular serial is handled by webpki and done with a naive linear scan. Some implementations (like BoringSSL) support maintaining a sorted list to allow faster search. We can probably rework the trait to support this but it will be trickier so for now I've left the simple solution in place. Pulled out to a separate issue: https://github.com/rustls/webpki/issues/80

* Similarly we perform CRL signature validation per-usage instead of once up-front. Pulled out to a separate issue: https://github.com/rustls/webpki/issues/81

* None of the revocation interface/checking is exposed for server cert validation, it's strictly for client cert validation. I think that's the most important use case and we can adapt the existing code for server cert chains pretty easily once stabilized.

Related to https://github.com/rustls/rustls/issues/1164, https://github.com/rustls/webpki/issues/57. Continues from https://github.com/rustls/webpki/pull/44.

### CODE REVIEW

File: src/verify_cert.rs
Comment (cpu) [inline src/verify_cert.rs:36]: The clippy error is related to the change to expose the `Cert` type.

I think I've come around to the idea of just sending the certificate serial & issuer through this trait but it's an open question whether we want to pass these as `untrusted::Input` and consider that part of our public API, or convert to some other representation to avoid that.

**Comment by cpu** [comment]
Rebased to resolve conflicts. Fixed clippy errs with some temp. rustdocs. Updated PR desc to reflect this only builds on 1 outstanding PR now (the CRL parsing pt1 work).

File: src/verify_cert.rs
Comment (ctz) [inline src/verify_cert.rs:238]: Latter sentence here is not adding much compared to the code.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs]: This requires exporting `EndEntityOrCa` but otherwise worked OK.

I lifted the depth enum out of the `wbepki` code and into the integration test driver `.rs` that's used by the generated test cases. I did this half to provide an example in case it was interesting for consumers, and half to keep the existing tests working more or less as-is. I'm also happy to simplify the tests to not handle EE vs CA distinctly.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs]: After paging this work back into my brain I'm remembering some of why this ends up being challenging. 

I think from a design perspective we don't want to push the signature validation to the user's `CrlProvider` implementation because it will be too easy to skip doing it, or doing it incorrectly so in my mind that option can be ruled out easily.

We wouldn't be able to do this with the `CertificateRevocationList::TryFrom` impl (since we wouldn't be able to provide an issuer argument in addition to the DER while meeting the trait's signature) so we would have to remove that trait to make progress on this front.

We could provide our own constructor that allowed specifying the issuer to use for validation, but we'd run into a complication from 5280. [RFC 5280 §6.3.3](https://www.rfc-editor.org/rfc/rfc5280#section-6.3.3) says:
>  (f)  Obtain and validate the certification path for the issuer of
>       the complete CRL.  The trust anchor for the certification
>       path MUST be the same as the trust anchor used to validate
>       the target certificate.

The MUST there will be tricky to enforce at CRL-load time when we won't know the trust anchor that will be used as the root for path-building until we actually build the path with the bundle of intermediates and trust anchors provided to `build_chain`. 

It also seems appealing to me that you could load a bunch of CRLs for end entities and intermediates without needing to source the intermediate issuers for validation ahead of time. If the CRL is relevant based on the intermediates used during path-building (likely sourced from the TLS handshake) they'll be used as required.

What do you think? Is there a middle ground I'm missing? :thinking:

**Comment by cpu** [comment]
>  cpu force-pushed the cpu-crl-support-pt2 branch from bb5d3da to 8425fa3 

Rebased, updated PR desc, and addressed the simple feedback.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs]: If we do end up wanting to do CRL signature validation at the time-of-load another challenge will be extracting and generalize the existing `build_chain` code to offer a way to verify a chain of signatures to a trust root without having an end entity certificate in-hand. 

The existing code is tightly coupled to the `Cert` representation and the `EndEntityOrCa` type that's used to build a linked list. We'd want to be able to express something like "CrlOrCa" instead. I don't love the idea of having two path building implementations to maintain, but I also struggled to generalize the existing one without making a mess. That also motivated my decision to consider revocation + the CRL signature validation as part of the existing `check_singatures` fn that's already walking the certificate linked list.

Comment (djc) [reply src/verify_cert.rs]: Not to distract from the consensus that we should postpone the optimization here, but it sounds like we should store the CRL and trust anchor together? That way we could assert in the type system that we've verified that the trust anchor has signed the CRL.

Comment (cpu) [reply src/verify_cert.rs]: > - perhaps we should just remember this as a future optimisation?

That sounds reasonable. I'll file an issue and leave a comment to point to it.

> it sounds like we should store the CRL and trust anchor together? That way we could assert in the type system that we've verified that the trust anchor has signed the CRL.

@djc Just to make sure I understand: you're suggesting that as a way to address the 5280 MUST concern in a world where we do the refactoring suggested by ctz, or is it a change you think should be incorporated with the current code?

**Comment by cpu** [comment]
> cpu force-pushed the cpu-crl-support-pt2 branch from 8425fa3 to 91c2fef 

Just some small non-functional changes (PR desc updated as well):
* removed a TODO I don't think is relevant anymore.
* added pointers to the optimization issues I filed. 

I also noticed I accidentally dropped the KeyUsage processing in one of my force pushes. Since this branch has a +1 from ctz and that work introduces new code across a few files I'm going to bring that back as a separate follow up. It's a change set that stands on its own so I think it will be easier to review in isolation from this larger work.

**Comment by cpu** [comment]
> I also noticed I accidentally dropped the KeyUsage processing in one of my force pushes. Since this branch has a +1 from ctz and that work introduces new code across a few files I'm going to bring that back as a separate follow up.

https://github.com/rustls/webpki/pull/82

File: src/verify_cert.rs
Comment (djc) [reply src/verify_cert.rs]: The former.

File: src/cert.rs
Comment (djc) [inline src/cert.rs:28]: Is the intent here that a `Cert` can only be created in webpki and passed to caller code as a parameter? If so, perhaps we should make this `#[non_exhaustive]` to avoid allowing downstream code to create their own `Cert` values?

Have we previously exposed `untrusted::Input<'_>` types in the public interface, and do we think that is the interface we want to expose to downstream callers?

File: src/end_entity.rs
Comment (djc) [inline src/end_entity.rs:121]: It is not obvious to me why we need the lifetime annotations here, given that we aren't returning any lifetime-bound data?

File: src/verify_cert.rs
Comment (djc) [inline src/verify_cert.rs:28]: If this is the reason for making `Cert` part of the public API, perhaps we can restrict the fields of `Cert` that are made public to a smaller subset of fields that we think are necessary to determine which CRL applies? Presumably that would be a smallish subset.

File: src/verify_cert.rs
Comment (djc) [inline src/verify_cert.rs:175]: Can we deduplicate the `check_crl()` calls here? They seem to be identical?

File: src/verify_cert.rs
Comment (djc) [inline src/verify_cert.rs:217]: Are we content with having a no-error interface here? I guess that might be sort of fundamental to CRLs, that not having a CRL gives the same result as getting the absence from a CRL? Do we think some callers might want to assert that a CRL was present for a trust anchor?

As an alternative, we could yield an `Option<CertificateUnrevoked>` token type, similar to the assertion types in rustls verification.

Style suggestions: I would inline the `crl_for_cert()` call and return the `None` case directly, as in

```rust
let crl = match crl_provider.crl_for_cert(cert) {
    Some(crl) => crl,
    None => return Ok(()),
};
```

File: src/verify_cert.rs
Comment (djc) [inline src/verify_cert.rs:229]: Style nit: no need for a `return` keyword here.

File: src/cert.rs
Comment (cpu) [reply src/cert.rs:28]: I waffled on both points a bit in some earlier [self-review](https://github.com/rustls/webpki/pull/66#discussion_r1196836037).

Looking through the existing exported types there's no `untrusted::Input` so I think it's clear we should avoid that.

Practically speaking that makes exporting the `Cert` type unhelpful since it's a big bag of `untrusted::Input`. With that in mind I think it's better to try to identify the most relevant fields and pass only those through to the CrlProvider. 

I'll rework :+1:

File: src/end_entity.rs
Comment (cpu) [reply src/end_entity.rs:121]: I'll take a look. I suspect its an artifact of some earlier iteration that I didn't clean up correctly.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs:28]: I agree (see my other comment).

I think we just need to send through:
* The serial (as `&[u8]`)
* The issuer (as `&[u8]`)
* An indication of whether the certificate is an end entity or intermediate (probably an enum made just for this purpose - the existing `EndEntityOrCa` enum exposes `Cert` in the `Ca` variant for the linked list)

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs:175]: Good catch, can do.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs:217]: > Are we content with having a no-error interface here? I guess that might be sort of fundamental to CRLs, that not having a CRL gives the same result as getting the absence from a CRL?

I guess I'm not sure what the webpki code will do with that distinction. If there's no CRL, there's no revocation status. If there's a CRL with no entry for the cert in question, there's no revocation status. In the end the flow of execution will continue the same way :thinking: 

> As an alternative, we could yield an Option<CertificateUnrevoked> token type, similar to the assertion types in rustls verification.

I'm open to this. Even though I don't think it will change how the webpki side operates I can see an argument that it could avoid bugs on the side of the trait implementer and lets us use the type system to maximize clarity of intent.

> Style suggestions: I would inline the crl_for_cert() call and return the None case directly, as in

Sounds good :+1:

File: src/cert.rs
Comment (djc) [reply src/cert.rs:28]: Could also still be the `Cert` type, but with methods instead of public fields?

File: src/verify_cert.rs
Comment (djc) [reply src/verify_cert.rs:217]: > I guess I'm not sure what the webpki code will do with that distinction. If there's no CRL, there's no revocation status. If there's a CRL with no entry for the cert in question, there's no revocation status. In the end the flow of execution will continue the same way 🤔

Yeah... I'm not completely sure myself. Just that in terms of control flow, it could help clarify what's going on -- it can be easy to just return `Ok(())` without paying much attention to what it means.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs]: :+1: Noted that on the tracking issue so it won't get lost when I resolve this comment: https://github.com/rustls/webpki/issues/81#issuecomment-1591587529

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs]: Fixed.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs]: > Style suggestions: I would inline the crl_for_cert() call and return the None case directly, as in

Done.

> As an alternative, we could yield an Option token type, similar to the assertion types in rustls verification.

Done. `check_crl` returns ` Result<Option<CertNotRevoked>, Error> ` now. We return `Ok(None)` when no CRL was found to use for revocation status, `Ok(Some(CertNotRevoked::assertion()))` when there was a CRL checked and the cert was found to be not revoked, and `Err(Error::CertRevoked)` when there was a CRL checked and the cert was found to be revoked.

File: src/verify_cert.rs
Comment (cpu) [reply src/verify_cert.rs]: Fixed.

File: src/cert.rs
Comment (cpu) [reply src/cert.rs:29]: > Could also still be the Cert type, but with methods instead of public fields? I don't see much point in inventing a wrapper type to expose to other code if we just need a higher-level abstraction here.

Reworked to handle as suggested.

Comment (cpu) [reply src/cert.rs:29]: > I'll rework that separately from this branch sweat

https://github.com/rustls/webpki/pull/83

I will resolve conflicts as these branches land.

**Comment by cpu** [comment]
>  cpu force-pushed the cpu-crl-support-pt2 branch from a232fdb to 3df6ae7 

I removed the key usage specific test suite from `generate.py` (and the generated test cases) so that I could pull them into https://github.com/rustls/webpki/pull/82 to better demonstrate the key usage enforcement.

File: src/end_entity.rs
Comment (djc) [reply src/end_entity.rs:121]: Nope, I guess it's fine.

**Review by djc — APPROVED** [review]

**Comment by cpu** [comment]
> cpu force-pushed the cpu-crl-support-pt2 branch from 3df6ae7 to 239cc88 

Rebasing to incorporate https://github.com/rustls/webpki/pull/83

I'm going to merge this when CI completes after checking with Ctz out-of-band that he's happy with this as-is.

**Pull request merged**
